crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -0,0 +1,117 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::PriceEstimationError,
+    anyhow::{anyhow, Result},
+    futures::{future::BoxFuture, FutureExt},
+    primitive_types::H160,
+    reqwest::{header::AUTHORIZATION, Client, StatusCode},
+    serde::Deserialize,
+    std::collections::HashMap,
+    url::Url,
+};
+
+#[derive(Debug, Deserialize)]
+struct Response(HashMap<H160, Price>);
+
+#[derive(Debug, Deserialize)]
+struct Price {
+    eth: f64,
+}
+
+type Token = H160;
+
+pub struct CoinGecko {
+    client: Client,
+    base_url: Url,
+    api_key: Option<String>,
+    chain: String,
+}
+
+impl CoinGecko {
+    pub fn new(
+        client: Client,
+        base_url: Url,
+        api_key: Option<String>,
+        chain_id: u64,
+    ) -> Result<Self> {
+        let chain = match chain_id {
+            1 => "ethereum".to_string(),
+            100 => "xdai".to_string(),
+            42161 => "arbitrum-one".to_string(),
+            n => anyhow::bail!("unknown network {n}"),
+        };
+        Ok(Self {
+            client,
+            base_url,
+            api_key,
+            chain,
+        })
+    }
+}
+
+impl NativePriceEstimating for CoinGecko {
+    fn estimate_native_price(&self, token: Token) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let mut builder = self.client.get(format!(
+                "{}/simple/token_price/{}?contract_addresses=0x{token:x}&vs_currencies=eth",
```

Comment:
I don't like this, as the base URL is not consistent, but we could have it temporarily. See the base URL is right now:

```
https://api.coingecko.com/api/v3
```

It is because that is the variable part compared to our BFF URL.

What's more, apparently BFF does not support `eth` prices yet, so I am forced to keep it like this so we can change the URL from official CoinGecko to BFF without modifying the code and just by adjusting the infrastructure.

Reply:
Technically, you could drop chain from the config here and pass the whole `https://api.coingecko.com/api/v3/simple/token_price/<chain>` in infra (and only add the get parameter in code). This way it wouldn't be inconsistent.

Reply:
good point! I will keep the `chain` as part of the `services` logic if you don't mind. Mostly to keep consistency with the other price estimators.